File: C++/ex22/hash_table_solution.cpp
/**
 * Exercise 22: Hash Table / Hash Map
 * Difficulty: **** (4/5)
 * 
 * Implement a simple hash table with collision resolution using linear probing.
 * Support insert, search, and delete operations.
 * 
 * Input:
 * - Commands: "insert key value", "search key", "delete key", or "exit"
 * 
 * Output:
 * - For search: print the value if found, or "not found"
 */

#include <bits/stdc++.h>
using namespace std;

const int TABLE_SIZE = 100;

struct HashEntry {
    string key;
    int value;
    bool is_deleted;
    
    HashEntry() : key(""), value(0), is_deleted(true) {}
    HashEntry(const string& k, int v) : key(k), value(v), is_deleted(false) {}
};

class HashTable {
private:
    vector<HashEntry> entries;
    
    unsigned int hashFunction(const string& key) {
        unsigned int hash = 0;
        for (char c : key) {
            hash = (hash << 5) + hash + c;
        }
        return hash % TABLE_SIZE;
    }
    
public:
    HashTable() : entries(TABLE_SIZE) {}
    
    void insert(const string& key, int value) {
        unsigned int index = hashFunction(key);
        
        while (!entries[index].is_deleted && entries[index].key != key) {
            index = (index + 1) % TABLE_SIZE;
        }
        
        entries[index] = HashEntry(key, value);
    }
    
    int search(const string& key) {
        unsigned int index = hashFunction(key);
        int iterations = 0;
        
        while (iterations < TABLE_SIZE) {
            if (!entries[index].is_deleted && entries[index].key == key) {
                return entries[index].value;
            }
            index = (index + 1) % TABLE_SIZE;
            iterations++;
        }
        
        return -1;
    }
    
    void deleteKey(const string& key) {
        unsigned int index = hashFunction(key);
        int iterations = 0;
        
        while (iterations < TABLE_SIZE) {
            if (!entries[index].is_deleted && entries[index].key == key) {
                entries[index].is_deleted = true;
                return;
            }
            index = (index + 1) % TABLE_SIZE;
            iterations++;
        }
    }
};

int main() {
    HashTable table;
    
    string cmd;
    while (cin >> cmd && cmd != "exit") {
        if (cmd == "insert") {
            string key;
            int value;
            cin >> key >> value;
            table.insert(key, value);
        } else if (cmd == "search") {
            string key;
            cin >> key;
            int result = table.search(key);
            if (result != -1) {
                cout << result << "\n";
            } else {
                cout << "not found\n";
            }
        } else if (cmd == "delete") {
            string key;
            cin >> key;
            table.deleteKey(key);
        }
    }
    
    return 0;
}

File: C#/ex22/hash_table_solution.cs
/**
 * Exercise 22: Hash Table / Hash Map
 * Difficulty: **** (4/5)
 * 
 * Implement a simple hash table with collision resolution using linear probing.
 * Support insert, search, and delete operations.
 * 
 * Input:
 * - Commands: "insert key value", "search key", "delete key", or "exit"
 * 
 * Output:
 * - For search: print the value if found, or "not found"
 */

using System;
using System.Collections.Generic;

class HashEntry {
    public string Key { get; set; }
    public int Value { get; set; }
    public bool IsDeleted { get; set; }
    
    public HashEntry(string key, int value) {
        Key = key;
        Value = value;
        IsDeleted = false;
    }
}

class HashTable {
    private const int TABLE_SIZE = 100;
    private HashEntry[] entries;
    
    public HashTable() {
        entries = new HashEntry[TABLE_SIZE];
        for (int i = 0; i < TABLE_SIZE; i++) {
            entries[i] = null;
        }
    }
    
    private int HashFunction(string key) {
        int hash = 0;
        foreach (char c in key) {
            hash = (hash << 5) + hash + c;
        }
        return Math.Abs(hash) % TABLE_SIZE;
    }
    
    public void Insert(string key, int value) {
        int index = HashFunction(key);
        
        while (entries[index] != null && !entries[index].IsDeleted) {
            if (entries[index].Key == key) {
                entries[index].Value = value;
                return;
            }
            index = (index + 1) % TABLE_SIZE;
        }
        
        entries[index] = new HashEntry(key, value);
    }
    
    public int Search(string key) {
        int index = HashFunction(key);
        int iterations = 0;
        
        while (entries[index] != null && iterations < TABLE_SIZE) {
            if (!entries[index].IsDeleted && entries[index].Key == key) {
                return entries[index].Value;
            }
            index = (index + 1) % TABLE_SIZE;
            iterations++;
        }
        
        return -1;
    }
    
    public void Delete(string key) {
        int index = HashFunction(key);
        int iterations = 0;
        
        while (entries[index] != null && iterations < TABLE_SIZE) {
            if (!entries[index].IsDeleted && entries[index].Key == key) {
                entries[index].IsDeleted = true;
                return;
            }
            index = (index + 1) % TABLE_SIZE;
            iterations++;
        }
    }
}

class Program {
    static void Main() {
        HashTable table = new HashTable();
        
        string cmd;
        while ((cmd = Console.ReadLine()) != null && cmd != "exit") {
            string[] parts = cmd.Split();
            
            if (parts[0] == "insert" && parts.Length == 3) {
                string key = parts[1];
                int value = int.Parse(parts[2]);
                table.Insert(key, value);
            } else if (parts[0] == "search" && parts.Length == 2) {
                string key = parts[1];
                int result = table.Search(key);
                if (result != -1) {
                    Console.WriteLine(result);
                } else {
                    Console.WriteLine("not found");
                }
            } else if (parts[0] == "delete" && parts.Length == 2) {
                string key = parts[1];
                table.Delete(key);
            }
        }
    }
}
